### ISSUE TITLE
Hand/finger pose estimation

### ISSUE THREAD

**Issue by carlosedubarreto** [zju3dv/EasyMocap#25]
Hi, I saw that openpose does this pose estimation, but, from what I saw, doesnt seems that keypose3d has this points.
Is is difficult to add there?

**Comment by chingswy** [comment]
No. It seems that I need the `fbx` model of `SMPL+H` or `SMPL-X` and modify the current export code. I have no time to it.

**Comment by carlosedubarreto** [comment]
@chingswy do you know i could output the fingers data on the SMPL json files?

From what I saw, when I run bodyhandface code, the output seems just a tiny big, from 72 it goes for 87 parameters, and reshaping it, it becomes 24 (if I do only the body estimation) to 29 (with bodyhandface)

It's just 5 more , and I think it's not enough to have the hand/fingers motion.

I'm going to search for that inside mp1f.py (I think that's the name) but if you already give me some hint where to find it, I'll go much faster.


Btw i already have the smplx model with hand and fingers, and I'm already tweaking the code to add them (at first I'm doing in MOCAP import, but later I can tweak also the script you did to export to bvh)

**Comment by chingswy** [comment]
> 
> It's just 5 more , and I think it's not enough to have the hand/fingers motion.
> 

The meaning of the poses parameters is:
```
22x3    +         6       +        6          +           3x3
body        left hand    right hand      jaw, left eye, right eye
```
Attention that the `6` for each hand doesn't represent its angles. To get the full poses, you can see this [func](https://github.com/zju3dv/EasyMocap/blob/master/easymocap/smplmodel/body_model.py#L135). I will add `--write_smpl_full` flag soon.

**Comment by carlosedubarreto** [comment]
thanks a lot, I'm going to dig it this week.

**Comment by carlosedubarreto** [comment]
I was analysing the code, and from what I saw, the same parameters I have at `smpl\000000.json` (and so on) are the same used by easymocap to create the render (with hands deformations)

Is that correct?

Because the rendered image looks very good, so, I think that amount of info:
```
22x3    +         6       +        6          +           3x3
body        left hand    right hand      jaw, left eye, right eye
```
is what is needed for me to use when importing it inside of blender.


I just wanted to ask you that, because if what I said is correct, then I'll stop analysing the code and will start working on how to make it work on the import to have a good representation as the render easymocap does.

**Comment by carlosedubarreto** [comment]
Just leaving it for reference, I could open the PKL from SMPLX for comparing where the joints are, since I didnt find it anywhere else.

```
'Pelvis': 0,
'L_Hip': 1,
'R_Hip': 2,
'Spine1': 3,
'L_Knee': 4,
'R_Knee': 5,
'Spine2': 6,
'L_Ankle': 7
'R_Ankle': 8,
'Spine3': 9,

'L_Foot': 10,
'R_Foot': 11,
'Neck': 12,
'L_Collar': 13,
'R_Collar': 14,
'Head': 15,
'L_Shoulder': 16,
'R_Shoulder': 17,
'L_Elbow': 18,
'R_Elbow': 19,

'L_Wrist': 20,
'R_Wrist': 21,
'Jaw': 22,
'L_Eye': 23,
'R_Eye': 24,
'L_Index1': 25,
'L_Index2': 26,
'L_Index3': 27,
'L_Middle1': 28,
'L_Middle2': 29,

'L_Middle3': 30,
'L_Pinky1': 31,
'L_Pinky2': 32,
'L_Pinky3': 33,
'L_Ring1': 34,
'L_Ring2': 35,
'L_Ring3': 36,
'L_Thumb1': 37,
'L_Thumb2': 38,
'L_Thumb3': 39,

'R_Index1': 40,
'R_Index2': 41,
'R_Index3': 42,
'R_Middle1': 43,
'R_Middle2': 44,
'R_Middle3': 45,
'R_Pinky1': 46,
'R_Pinky2': 47,
'R_Pinky3': 48,
'R_Ring1': 49,

'R_Ring2': 50,
'R_Ring3': 51,
'R_Thumb1': 52,
'R_Thumb2': 53,
'R_Thumb3': 54,
```

to open and see this I used a tip from this issue
https://github.com/facebookresearch/DensePose/issues/142#issuecomment-435645312

that is to use this sort of code:
```
import pickle
import gzip
import numpy
with open('mnist.pkl', 'rb') as f:
    u = pickle._Unpickler(f)
    u.encoding = 'latin1'
    p = u.load()
    print(p)
```

**Comment by carlosedubarreto** [comment]
Hi @chingswy 
I spend most of my day trying to get how to map the parameters using the data in SMPL.
I think I got, but even though, I was not sucessfull on getting something working.

the I tried to debug easymocap to understand how it can work on its on render (making the render of the hand /fingers correcly) I was trying to get the same data that easymocap renders.

Then I noticed that it looks like I already have almost the same data.... But I saw one problem. the precision of the exported number, I think that´s what I'm having problems with.

take a look

![image](https://user-images.githubusercontent.com/4061130/118390870-cd47e780-b607-11eb-974e-48f91602eca6.png)

I can see that it tries to export the data to the fingers, but the values are so low, that on the json output, it only comes zeros.

I'll try to increase the precision to see if I get any better result.

**Comment by carlosedubarreto** [comment]
I increased the size to 10, and now some usefull numbers show.
I'll test it and let you know.
![Code_Sw0xRP35EY](https://user-images.githubusercontent.com/4061130/118391147-2a906880-b609-11eb-97e8-fd63e2cb8a7c.png)


![image](https://user-images.githubusercontent.com/4061130/118391143-2401f100-b609-11eb-9481-bfd168805ce4.png)

**Comment by carlosedubarreto** [comment]
Almost a week later I could understand your message below

 get the full poses, you can see this [func](https://github.com/zju3dv/EasyMocap/blob/master/easymocap/smplmodel/body_model.py#L135). I will add `--write_smpl_full` flag soon.

And I only understood it after debugging the code. There I have all the 165 values I need to create to poses outside easymocap.
I'm trying to export it from there to use on my addon.

Thanks a lot for the information. Ill pay more attention on the next time 😅

**Comment by carlosedubarreto** [comment]
at the end I fixed my problem doing the easy way out LOL. I forced the output of the parameters on body_model, and then copied it to the end folder

like this
exporting the data with a fixed name
![image](https://user-images.githubusercontent.com/4061130/118484740-f478e480-b6ed-11eb-93ef-2c9887ff3a80.png)

and copied it to the end folder in here
![image](https://user-images.githubusercontent.com/4061130/118484876-196d5780-b6ee-11eb-8729-59541ffa4b92.png)

Its working very well. I'm loving it. thanks for all the help!!

**Comment by carlosedubarreto** [comment]
Sample of the result.

[https://youtu.be/aCw_yS0kEEM](https://youtu.be/aCw_yS0kEEM)

**Comment by chingswy** [comment]
Hello @carlosedubarreto , you can add a simple flag `--write_smpl_full` now to get the full poses.

**Comment by carlosedubarreto** [comment]
Thanks a lot!!!!!!

**Comment by carlosedubarreto** [comment]
> Hello @carlosedubarreto , you can add a simple flag `--write_smpl_full` now to get the full poses.

Hi @chingswy 
I downloaded the new version of easymocap and tried to run this new option but where should the result appear?

I thought it was going to show on the smpl json files, but the pose has only 87 values.

I ran the command like this
`python apps/demo/mv1p.py 0_input/sample --out 1_output/sample/smplx3 --undis --sub_vis 1 7 13 19 --body bodyhandface --model smplx --gender male --write_smpl_full`

**Comment by chingswy** [comment]
The results will be stored at `${out}/smpl_full`

**Comment by carlosedubarreto** [comment]
I think it is not creating the folder 😅
![image](https://user-images.githubusercontent.com/4061130/122525000-27650f80-cfef-11eb-866b-b5511f2559b2.png)

**Comment by chingswy** [comment]
Forget to commit it. Please try it again.

**Comment by carlosedubarreto** [comment]
Great!!!!. thanks. Its working now....

I'm gonig to update the addon to use this....

Thanks a lot!!!

**Comment by carlosedubarreto** [comment]
Here is the updated version.

You just have to run easymocap adding `--write_smpl_full`

for example
`python apps/demo/mv1p.py 0_input/sample --out 1_output/sample/smplx --vis_det --vis_repro --undis --sub_vis 1 7 13 19 --body bodyhandface --model smplx --gender male --vis_smpl --write_smpl_full`

[Mocap_PE_data_import_0.722.zip](https://github.com/zju3dv/EasyMocap/files/6675770/Mocap_PE_data_import_0.722.zip)


when using the addon, you choose the folder SMPL_FULL

To have the latest version of the blender addon, follow these links.


patreon.com/cebstudios

https://www.youtube.com/c/CEBStudios

https://github.com/carlosedubarreto/b3d_mocap_import

https://gumroad.com/l/mocap_import

https://twitter.com/carlosedubarret
